src/cplscheme/BaseCouplingScheme.cpp

```diff
@@ -188,14 +161,13 @@ std::vector<int> BaseCouplingScheme::receiveData(
   std::vector<int> receivedDataIDs;
   PRECICE_ASSERT(m2n.get() != nullptr);
   PRECICE_ASSERT(m2n->isConnected());
-
   for (DataMap::value_type &pair : _receiveData) {
     int size = pair.second->values->size();
     m2n->receive(pair.second->values->data(), size, pair.second->mesh->getID(), pair.second->dimension);
     receivedDataIDs.push_back(pair.first);
   }
   PRECICE_DEBUG("Number of received data sets = " << receivedDataIDs.size());
-
+  _hasDataBeenExchanged = true;
```

Comment:
The usage of `_hasDataBeenExchanged` looks a bit mixed up.

Reply:
I agree. `_hasDataBeenExchanged` is a flag that is used at several places. The idea is the following: set `_hasDataBeenExchanged = false` at the beginning of a function that has to exchange data (`advance` and `initializeData`). As soon as data has been exchanged successfully, `_hasDataBeenExchanged` is set to `true`. This happens in different context (`receiveData` and `updateTimeAndIterations`), which makes the situation a bit confusing.

I tried to improve the situation in the following way:

* remove `_hasDataBeenExchanged = true` from `updateTimeAndIterations`, since it was not there before. Probably slipped in during refactoring.
* make sure that flag is not set `true`, if it is already `true`
* change the behaviour and name of the setter. public function should only be used to set flag. `BaseCouplingScheme` internally unsets flag.
* make name more specific: `_hasDataBeenReceived`

However, this led to a situation where the use of `_cplscheme.hasDataBeenReceived()` by the tests and `SolverInterfaceImpl` did not really make sense. Therefore, I reverted my changes. I think we should agree on one of the following solutions:

1) have a single flag `_hasDataBeenExchanged` that is set by the coupling scheme at the right place (not inside `receiveData`, since it refers to data exchange = send and receive).
2) have two flags `_hasDataBeenSent` and `_hasDataBeenReceived` that are set by coupling scheme. The function call `_cplscheme.hasDataBeenExchanged()` evaluates these two flags.

Reply:
I vote for solution 1.